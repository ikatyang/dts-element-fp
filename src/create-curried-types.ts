import * as dts from 'dts-element';
import * as R from 'ramda';
import {
  get_function_parameter_placeholder_name_default,
  get_function_type_name_default,
  get_placeholder_type_default,
  inline_return_type_default,
  placeholder_default,
  selectable_default,
} from './utils/constants';
import { create_masks } from './utils/create-masks';
import {
  create_selectable_signatures,
  CreateSelectableSignaturesOptions,
} from './utils/create-selectable-signatures';
import { get_generics_dependencies } from './utils/get-generics-dependencies';
import { has } from './utils/has';
import { push_signatures } from './utils/push-signatures';

export interface CreateCurriedDeclarationsOptions
  extends CreateSelectableSignaturesOptions {
  selectable?: boolean;
  placeholder?: boolean;
  get_placeholder_type?: () => string;
  get_function_type_name?: (name: string, mask: string) => string;
  get_function_parameter_placeholder_name?: (name: string) => string;
  inline_return_type?: boolean;
}

/**
 * @hidden
 */
export function create_curried_types(
  name: string,
  type: dts.IFunctionType,
  // istanbul ignore next
  options: CreateCurriedDeclarationsOptions = {},
): dts.ITypeDeclaration[] {
  // istanbul ignore next
  const {
    selectable = selectable_default,
    placeholder = placeholder_default,
    get_placeholder_type = get_placeholder_type_default,
    get_function_type_name = get_function_type_name_default,
    get_function_parameter_placeholder_name = get_function_parameter_placeholder_name_default,
    get_selectable_kind_name,
    get_selectable_selector_name,
    inline_return_type = inline_return_type_default,
  } = options;

  const is_placeholder = (value: dts.IType) =>
    dts.is_general_type(value) && value.name === get_placeholder_type();

  const { generics = [], parameters = [], return: return_type } = type;

  const type_predicate_parameter = dts.is_type_predicate(return_type)
    ? parameters.find(
        parameter => parameter.name === (return_type.parameter as string),
      )!
    : null;

  const placeholders = parameters.map(parameter =>
    dts.create_parameter_declaration({
      name: get_function_parameter_placeholder_name(parameter.name),
      type: dts.create_general_type({ name: get_placeholder_type() }),
    }),
  );

  const generics_dependencies = get_generics_dependencies(generics);

  const sort_generics = (the_generics: dts.IGenericDeclaration[]) =>
    the_generics
      .slice()
      .sort((a, b) => generics.indexOf(a) - generics.indexOf(b));

  const add_generic_dependencies = (
    the_generics: dts.IGenericDeclaration[],
  ) => {
    const added: dts.IGenericDeclaration[] = [];

    the_generics.forEach(generic => {
      added.push(...generics_dependencies[generics.indexOf(generic)]);
    });

    return sort_generics(R.uniq(added));
  };

  const parameters_generics = parameters.map(parameter =>
    add_generic_dependencies(
      generics.filter(generic =>
        has(parameter, {
          kind: dts.ElementKind.GeneralType,
          name: generic.name,
        }),
      ),
    ),
  );

  const return_generics = add_generic_dependencies(
    generics.filter(generic =>
      has(return_type, {
        kind: dts.ElementKind.GeneralType,
        name: generic.name,
      }),
    ),
  );

  const target_types = R.repeat(0, 2 ** parameters.length).map(
    (_, index, array) =>
      index === array.length - 1
        ? dts.is_type_predicate(return_type) ? dts.boolean_type : return_type
        : dts.create_object_type({ members: [] }),
  );

  const masks = create_masks(parameters.length);
  const reverse_masks = masks.reduce<{ [mask: string]: number }>(
    (current_reverse_masks, mask, index) => ({
      ...current_reverse_masks,
      [mask]: index,
    }),
    {},
  );

  const type_declarations = target_types.map((target_type, index) => {
    const type_generics = sort_generics(
      index === target_types.length - 1
        ? return_generics
        : parameters_generics
            .filter((_, generics_index) => masks[index][generics_index] === '1')
            .reduce(
              (current_type_generics, filtered_parameter_generics) =>
                R.uniq(
                  current_type_generics.concat(filtered_parameter_generics),
                ),
              [],
            ),
    );
    return dts.create_type_declaration({
      name: get_function_type_name(name, masks[index]),
      generics: type_generics,
      type: target_type,
    });
  });

  type_declarations.slice(0, -1).forEach((type_declaration, index) => {
    const members = (type_declaration.type as dts.IObjectType)
      .members as dts.IObjectMember[];
    push_signatures(
      is_placeholder,
      parameters.filter((_, param_index) => masks[index][param_index] === '0'),
      placeholder
        ? placeholders.filter(
            (_, param_index) => masks[index][param_index] === '0',
          )
        : [],
      (used_parameters, return_parameters) => {
        const return_mask = parameters
          .map(
            parameter =>
              return_parameters.indexOf(parameter) === -1 ? '1' : '0',
          )
          .join('');
        const return_type_declaration =
          type_declarations[reverse_masks[return_mask]];
        const return_type_generics = return_type_declaration.generics!.map(
          generic => dts.create_general_type({ name: generic.name }),
        );
        const used_parameters_generics = sort_generics(
          R.uniq(
            used_parameters.reduce(
              (current, parameter) => [
                ...current,
                ...(parameter_index =>
                  parameter_index === -1
                    ? []
                    : parameters_generics[parameter_index])(
                  parameters.indexOf(parameter),
                ),
              ],
              return_type_declaration.generics!,
            ),
          ),
        ).filter(generic => type_declaration.generics!.indexOf(generic) === -1);
        members.push(
          dts.create_object_member({
            owned: dts.create_function_declaration({
              name: undefined,
              type: dts.create_function_type({
                generics: used_parameters_generics,
                parameters: used_parameters,
                return: get_overload_return_type(),
              }),
            }),
          }),
        );

        function get_overload_return_type() {
          if (return_mask.split('').every(R.equals('1'))) {
            if (type_predicate_parameter !== null) {
              if (used_parameters.indexOf(type_predicate_parameter) !== -1) {
                return return_type; // type_predicate: value is Type (matched parameter)
              }
              if (inline_return_type) {
                return dts.boolean_type; // type_predicate: boolean (unmatched parameter)
              }
            } else if (inline_return_type) {
              return return_type;
            }
          }
          return dts.create_general_type({
            name: return_type_declaration.name,
            generics: return_type_generics,
          });
        }
      },
    );

    if (!selectable || members.length <= 1) {
      return;
    }

    members.splice(
      -1,
      0,
      ...create_selectable_signatures(members, {
        is_placeholder,
        get_selectable_kind_name,
        get_selectable_selector_name,
      }),
    );
  });

  return type_declarations;
}
