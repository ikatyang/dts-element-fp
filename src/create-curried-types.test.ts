import * as dts from 'dts-element';
import { create_curried_types } from './create-curried-types';

const test_cases: Record<string, string> = {
  'non-generics': `
    function clamp(min: number, max: number, value: number): number;
  `,
  'generics + multi-parameters': `
    function all<T>(fn: Predicate<T>, list: List<T>): boolean;
  `,
  'generics + multi-parameters + return-generics': `
    function adjust<T, U>(fn: Morphism<T, U>, index: number, list: List<T>): (T | U)[];
  `,
  'type-predicate': `
    function is<T>(constructor: Constructor<T>, value: any): value is T;
  `,
};

Object.keys(test_cases).forEach(case_name => {
  describe(case_name, () => {
    it('should transform correctly without any options', () => {
      expect(emit_curried_types()).toMatchSnapshot();
    });

    it('should transform correctly with selectable', () => {
      expect(
        emit_curried_types({
          selectable: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform correctly with placeholder', () => {
      expect(
        emit_curried_types({
          placeholder: true,
        }),
      ).toMatchSnapshot();
    });

    it('should transform correctly with every option', () => {
      expect(
        emit_curried_types({
          selectable: true,
          placeholder: true,
        }),
      ).toMatchSnapshot();
    });
  });

  function emit_curried_types(
    options: { selectable?: boolean; placeholder?: boolean } = {},
  ) {
    const test_case = test_cases[case_name];
    const function_declaration = dts.parse(test_case)
      .members[0] as dts.IFunctionDeclaration;

    const curried_types = create_curried_types(
      function_declaration.name!,
      function_declaration.type!,
      options,
    );
    const top_level_element = dts.create_top_level_element({
      members: curried_types,
    });
    return dts.emit(top_level_element);
  }
});
