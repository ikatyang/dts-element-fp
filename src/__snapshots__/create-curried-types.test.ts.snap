// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generics + multi-parameters + return-generics should transform correctly with every option 1`] = `
"type adjust_000 = {
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number): adjust_110<T, U>;
    <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>): adjust_101<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_100<T, U>;
    <T>(_fn: PH, index: number, list: List<T>): adjust_011<T>;
    (_fn: PH, index: number): adjust_010;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => adjust_110<T, U>;
    <$SEL extends \\"101\\">(): <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>) => adjust_101<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => adjust_100<T, U>;
    <$SEL extends \\"011\\">(): <T>(_fn: PH, index: number, list: List<T>) => adjust_011<T>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => adjust_010;
    <$SEL extends \\"001\\">(): <T>(_fn: PH, _index: PH, list: List<T>) => adjust_001<T>;
    <T>(_fn: PH, _index: PH, list: List<T>): adjust_001<T>;
};
type adjust_100<T, U> = {
    (index: number, list: List<T>): adjust_111<T, U>;
    (index: number): adjust_110<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => adjust_110<T, U>;
    <$SEL extends \\"01\\">(): (_index: PH, list: List<T>) => adjust_101<T, U>;
    (_index: PH, list: List<T>): adjust_101<T, U>;
};
type adjust_010 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_110<T, U>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => adjust_110<T, U>;
    <$SEL extends \\"01\\">(): <T>(_fn: PH, list: List<T>) => adjust_011<T>;
    <T>(_fn: PH, list: List<T>): adjust_011<T>;
};
type adjust_110<T, U> = {
    (list: List<T>): adjust_111<T, U>;
};
type adjust_001<T> = {
    <U>(fn: Morphism<T, U>, index: number): adjust_111<T, U>;
    <U>(fn: Morphism<T, U>): adjust_101<T, U>;
    <$SEL extends \\"11\\">(): <U>(fn: Morphism<T, U>, index: number) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): <U>(fn: Morphism<T, U>) => adjust_101<T, U>;
    <$SEL extends \\"01\\">(): (_fn: PH, index: number) => adjust_011<T>;
    (_fn: PH, index: number): adjust_011<T>;
};
type adjust_101<T, U> = {
    (index: number): adjust_111<T, U>;
};
type adjust_011<T> = {
    <U>(fn: Morphism<T, U>): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with placeholder 1`] = `
"type adjust_000 = {
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number): adjust_110<T, U>;
    <T, U>(fn: Morphism<T, U>, _index: PH, list: List<T>): adjust_101<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_100<T, U>;
    <T>(_fn: PH, index: number, list: List<T>): adjust_011<T>;
    (_fn: PH, index: number): adjust_010;
    <T>(_fn: PH, _index: PH, list: List<T>): adjust_001<T>;
};
type adjust_100<T, U> = {
    (index: number, list: List<T>): adjust_111<T, U>;
    (index: number): adjust_110<T, U>;
    (_index: PH, list: List<T>): adjust_101<T, U>;
};
type adjust_010 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_110<T, U>;
    <T>(_fn: PH, list: List<T>): adjust_011<T>;
};
type adjust_110<T, U> = {
    (list: List<T>): adjust_111<T, U>;
};
type adjust_001<T> = {
    <U>(fn: Morphism<T, U>, index: number): adjust_111<T, U>;
    <U>(fn: Morphism<T, U>): adjust_101<T, U>;
    (_fn: PH, index: number): adjust_011<T>;
};
type adjust_101<T, U> = {
    (index: number): adjust_111<T, U>;
};
type adjust_011<T> = {
    <U>(fn: Morphism<T, U>): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly with selectable 1`] = `
"type adjust_000 = {
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number): adjust_110<T, U>;
    <$SEL extends \\"111\\">(): <T, U>(fn: Morphism<T, U>, index: number, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, index: number) => adjust_110<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => adjust_100<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_100<T, U>;
};
type adjust_100<T, U> = {
    (index: number, list: List<T>): adjust_111<T, U>;
    <$SEL extends \\"11\\">(): (index: number, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): (index: number) => adjust_110<T, U>;
    (index: number): adjust_110<T, U>;
};
type adjust_010 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): adjust_111<T, U>;
    <$SEL extends \\"11\\">(): <T, U>(fn: Morphism<T, U>, list: List<T>) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): <T, U>(fn: Morphism<T, U>) => adjust_110<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_110<T, U>;
};
type adjust_110<T, U> = {
    (list: List<T>): adjust_111<T, U>;
};
type adjust_001<T> = {
    <U>(fn: Morphism<T, U>, index: number): adjust_111<T, U>;
    <$SEL extends \\"11\\">(): <U>(fn: Morphism<T, U>, index: number) => adjust_111<T, U>;
    <$SEL extends \\"1\\">(): <U>(fn: Morphism<T, U>) => adjust_101<T, U>;
    <U>(fn: Morphism<T, U>): adjust_101<T, U>;
};
type adjust_101<T, U> = {
    (index: number): adjust_111<T, U>;
};
type adjust_011<T> = {
    <U>(fn: Morphism<T, U>): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];
"
`;

exports[`generics + multi-parameters + return-generics should transform correctly without any options 1`] = `
"type adjust_000 = {
    <T, U>(fn: Morphism<T, U>, index: number, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>, index: number): adjust_110<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_100<T, U>;
};
type adjust_100<T, U> = {
    (index: number, list: List<T>): adjust_111<T, U>;
    (index: number): adjust_110<T, U>;
};
type adjust_010 = {
    <T, U>(fn: Morphism<T, U>, list: List<T>): adjust_111<T, U>;
    <T, U>(fn: Morphism<T, U>): adjust_110<T, U>;
};
type adjust_110<T, U> = {
    (list: List<T>): adjust_111<T, U>;
};
type adjust_001<T> = {
    <U>(fn: Morphism<T, U>, index: number): adjust_111<T, U>;
    <U>(fn: Morphism<T, U>): adjust_101<T, U>;
};
type adjust_101<T, U> = {
    (index: number): adjust_111<T, U>;
};
type adjust_011<T> = {
    <U>(fn: Morphism<T, U>): adjust_111<T, U>;
};
type adjust_111<T, U> = (T | U)[];
"
`;

exports[`generics + multi-parameters should transform correctly with every option 1`] = `
"type all_00 = {
    <T>(fn: Predicate<T>, list: List<T>): all_11;
    <T>(fn: Predicate<T>): all_10<T>;
    <$SEL extends \\"11\\">(): <T>(fn: Predicate<T>, list: List<T>) => all_11;
    <$SEL extends \\"1\\">(): <T>(fn: Predicate<T>) => all_10<T>;
    <$SEL extends \\"01\\">(): <T>(_fn: PH, list: List<T>) => all_01<T>;
    <T>(_fn: PH, list: List<T>): all_01<T>;
};
type all_10<T> = {
    (list: List<T>): all_11;
};
type all_01<T> = {
    (fn: Predicate<T>): all_11;
};
type all_11 = boolean;
"
`;

exports[`generics + multi-parameters should transform correctly with placeholder 1`] = `
"type all_00 = {
    <T>(fn: Predicate<T>, list: List<T>): all_11;
    <T>(fn: Predicate<T>): all_10<T>;
    <T>(_fn: PH, list: List<T>): all_01<T>;
};
type all_10<T> = {
    (list: List<T>): all_11;
};
type all_01<T> = {
    (fn: Predicate<T>): all_11;
};
type all_11 = boolean;
"
`;

exports[`generics + multi-parameters should transform correctly with selectable 1`] = `
"type all_00 = {
    <T>(fn: Predicate<T>, list: List<T>): all_11;
    <$SEL extends \\"11\\">(): <T>(fn: Predicate<T>, list: List<T>) => all_11;
    <$SEL extends \\"1\\">(): <T>(fn: Predicate<T>) => all_10<T>;
    <T>(fn: Predicate<T>): all_10<T>;
};
type all_10<T> = {
    (list: List<T>): all_11;
};
type all_01<T> = {
    (fn: Predicate<T>): all_11;
};
type all_11 = boolean;
"
`;

exports[`generics + multi-parameters should transform correctly without any options 1`] = `
"type all_00 = {
    <T>(fn: Predicate<T>, list: List<T>): all_11;
    <T>(fn: Predicate<T>): all_10<T>;
};
type all_10<T> = {
    (list: List<T>): all_11;
};
type all_01<T> = {
    (fn: Predicate<T>): all_11;
};
type all_11 = boolean;
"
`;

exports[`non-generics should transform correctly with every option 1`] = `
"type clamp_000 = {
    (min: number, max: number, value: number): clamp_111;
    (min: number, max: number): clamp_110;
    (min: number, _max: PH, value: number): clamp_101;
    (min: number): clamp_100;
    (_min: PH, max: number, value: number): clamp_011;
    (_min: PH, max: number): clamp_010;
    <$SEL extends \\"111\\">(): (min: number, max: number, value: number) => clamp_111;
    <$SEL extends \\"11\\">(): (min: number, max: number) => clamp_110;
    <$SEL extends \\"101\\">(): (min: number, _max: PH, value: number) => clamp_101;
    <$SEL extends \\"1\\">(): (min: number) => clamp_100;
    <$SEL extends \\"011\\">(): (_min: PH, max: number, value: number) => clamp_011;
    <$SEL extends \\"01\\">(): (_min: PH, max: number) => clamp_010;
    <$SEL extends \\"001\\">(): (_min: PH, _max: PH, value: number) => clamp_001;
    (_min: PH, _max: PH, value: number): clamp_001;
};
type clamp_100 = {
    (max: number, value: number): clamp_111;
    (max: number): clamp_110;
    <$SEL extends \\"11\\">(): (max: number, value: number) => clamp_111;
    <$SEL extends \\"1\\">(): (max: number) => clamp_110;
    <$SEL extends \\"01\\">(): (_max: PH, value: number) => clamp_101;
    (_max: PH, value: number): clamp_101;
};
type clamp_010 = {
    (min: number, value: number): clamp_111;
    (min: number): clamp_110;
    <$SEL extends \\"11\\">(): (min: number, value: number) => clamp_111;
    <$SEL extends \\"1\\">(): (min: number) => clamp_110;
    <$SEL extends \\"01\\">(): (_min: PH, value: number) => clamp_011;
    (_min: PH, value: number): clamp_011;
};
type clamp_110 = {
    (value: number): clamp_111;
};
type clamp_001 = {
    (min: number, max: number): clamp_111;
    (min: number): clamp_101;
    <$SEL extends \\"11\\">(): (min: number, max: number) => clamp_111;
    <$SEL extends \\"1\\">(): (min: number) => clamp_101;
    <$SEL extends \\"01\\">(): (_min: PH, max: number) => clamp_011;
    (_min: PH, max: number): clamp_011;
};
type clamp_101 = {
    (max: number): clamp_111;
};
type clamp_011 = {
    (min: number): clamp_111;
};
type clamp_111 = number;
"
`;

exports[`non-generics should transform correctly with placeholder 1`] = `
"type clamp_000 = {
    (min: number, max: number, value: number): clamp_111;
    (min: number, max: number): clamp_110;
    (min: number, _max: PH, value: number): clamp_101;
    (min: number): clamp_100;
    (_min: PH, max: number, value: number): clamp_011;
    (_min: PH, max: number): clamp_010;
    (_min: PH, _max: PH, value: number): clamp_001;
};
type clamp_100 = {
    (max: number, value: number): clamp_111;
    (max: number): clamp_110;
    (_max: PH, value: number): clamp_101;
};
type clamp_010 = {
    (min: number, value: number): clamp_111;
    (min: number): clamp_110;
    (_min: PH, value: number): clamp_011;
};
type clamp_110 = {
    (value: number): clamp_111;
};
type clamp_001 = {
    (min: number, max: number): clamp_111;
    (min: number): clamp_101;
    (_min: PH, max: number): clamp_011;
};
type clamp_101 = {
    (max: number): clamp_111;
};
type clamp_011 = {
    (min: number): clamp_111;
};
type clamp_111 = number;
"
`;

exports[`non-generics should transform correctly with selectable 1`] = `
"type clamp_000 = {
    (min: number, max: number, value: number): clamp_111;
    (min: number, max: number): clamp_110;
    <$SEL extends \\"111\\">(): (min: number, max: number, value: number) => clamp_111;
    <$SEL extends \\"11\\">(): (min: number, max: number) => clamp_110;
    <$SEL extends \\"1\\">(): (min: number) => clamp_100;
    (min: number): clamp_100;
};
type clamp_100 = {
    (max: number, value: number): clamp_111;
    <$SEL extends \\"11\\">(): (max: number, value: number) => clamp_111;
    <$SEL extends \\"1\\">(): (max: number) => clamp_110;
    (max: number): clamp_110;
};
type clamp_010 = {
    (min: number, value: number): clamp_111;
    <$SEL extends \\"11\\">(): (min: number, value: number) => clamp_111;
    <$SEL extends \\"1\\">(): (min: number) => clamp_110;
    (min: number): clamp_110;
};
type clamp_110 = {
    (value: number): clamp_111;
};
type clamp_001 = {
    (min: number, max: number): clamp_111;
    <$SEL extends \\"11\\">(): (min: number, max: number) => clamp_111;
    <$SEL extends \\"1\\">(): (min: number) => clamp_101;
    (min: number): clamp_101;
};
type clamp_101 = {
    (max: number): clamp_111;
};
type clamp_011 = {
    (min: number): clamp_111;
};
type clamp_111 = number;
"
`;

exports[`non-generics should transform correctly without any options 1`] = `
"type clamp_000 = {
    (min: number, max: number, value: number): clamp_111;
    (min: number, max: number): clamp_110;
    (min: number): clamp_100;
};
type clamp_100 = {
    (max: number, value: number): clamp_111;
    (max: number): clamp_110;
};
type clamp_010 = {
    (min: number, value: number): clamp_111;
    (min: number): clamp_110;
};
type clamp_110 = {
    (value: number): clamp_111;
};
type clamp_001 = {
    (min: number, max: number): clamp_111;
    (min: number): clamp_101;
};
type clamp_101 = {
    (max: number): clamp_111;
};
type clamp_011 = {
    (min: number): clamp_111;
};
type clamp_111 = number;
"
`;

exports[`type-predicate should transform correctly with every option 1`] = `
"type is_00 = {
    <T>(constructor: Constructor<T>, value: any): value is T;
    <T>(constructor: Constructor<T>): is_10<T>;
    <$SEL extends \\"11\\">(): <T>(constructor: Constructor<T>, value: any) => value is T;
    <$SEL extends \\"1\\">(): <T>(constructor: Constructor<T>) => is_10<T>;
    <$SEL extends \\"01\\">(): (_constructor: PH, value: any) => is_01;
    (_constructor: PH, value: any): is_01;
};
type is_10<T> = {
    (value: any): value is T;
};
type is_01 = {
    <T>(constructor: Constructor<T>): is_11<T>;
};
type is_11<T> = boolean;
"
`;

exports[`type-predicate should transform correctly with placeholder 1`] = `
"type is_00 = {
    <T>(constructor: Constructor<T>, value: any): value is T;
    <T>(constructor: Constructor<T>): is_10<T>;
    (_constructor: PH, value: any): is_01;
};
type is_10<T> = {
    (value: any): value is T;
};
type is_01 = {
    <T>(constructor: Constructor<T>): is_11<T>;
};
type is_11<T> = boolean;
"
`;

exports[`type-predicate should transform correctly with selectable 1`] = `
"type is_00 = {
    <T>(constructor: Constructor<T>, value: any): value is T;
    <$SEL extends \\"11\\">(): <T>(constructor: Constructor<T>, value: any) => value is T;
    <$SEL extends \\"1\\">(): <T>(constructor: Constructor<T>) => is_10<T>;
    <T>(constructor: Constructor<T>): is_10<T>;
};
type is_10<T> = {
    (value: any): value is T;
};
type is_01 = {
    <T>(constructor: Constructor<T>): is_11<T>;
};
type is_11<T> = boolean;
"
`;

exports[`type-predicate should transform correctly without any options 1`] = `
"type is_00 = {
    <T>(constructor: Constructor<T>, value: any): value is T;
    <T>(constructor: Constructor<T>): is_10<T>;
};
type is_10<T> = {
    (value: any): value is T;
};
type is_01 = {
    <T>(constructor: Constructor<T>): is_11<T>;
};
type is_11<T> = boolean;
"
`;
